cudax/include/cuda/experimental/__container/async_buffer.cuh

```diff
@@ -0,0 +1,858 @@
+  using iterator               = heterogeneous_iterator<_Tp, false, _Properties...>;
+  using const_iterator         = heterogeneous_iterator<_Tp, true, _Properties...>;
+  using reverse_iterator       = _CUDA_VSTD::reverse_iterator<iterator>;
+  using const_reverse_iterator = _CUDA_VSTD::reverse_iterator<const_iterator>;
+  using size_type              = _CUDA_VSTD::size_t;
+  using difference_type        = _CUDA_VSTD::ptrdiff_t;
+
+  using __env_t          = ::cuda::experimental::env_t<_Properties...>;
+  using __policy_t       = ::cuda::experimental::execution::execution_policy;
+  using __buffer_t       = ::cuda::experimental::uninitialized_async_buffer<_Tp, _Properties...>;
+  using __resource_t     = ::cuda::experimental::any_async_resource<_Properties...>;
+  using __resource_ref_t = _CUDA_VMR::async_resource_ref<_Properties...>;
+
+  template <class, class...>
+  friend class async_buffer;
+
+  // For now we require trivially copyable type to simplify the implementation
+  static_assert(_CCCL_TRAIT(_CUDA_VSTD::is_trivially_copyable, _Tp),
+                "cuda::experimental::async_buffer requires T to be trivially copyable.");
+
+  // At least one of the properties must signal an execution space
+  static_assert(_CUDA_VMR::__contains_execution_space_property<_Properties...>,
+                "The properties of cuda::experimental::async_buffer must contain at least one execution space "
+                "property!");
+
+  //! @brief Convenience shortcut to detect the execution space of the async_buffer
+  static constexpr bool __is_host_only = __select_execution_space<_Properties...> == _ExecutionSpace::__host;
+
+private:
+  __buffer_t __buf_;
+  size_type __size_    = 0; // initialized to 0 in case initialization of the elements might throw
+  __policy_t __policy_ = __policy_t::invalid_execution_policy;
+
+  //! @brief Helper to check container is compatible with this async_buffer
+  template <class _Range>
+  static constexpr bool __compatible_range = _CUDA_VRANGES::__container_compatible_range<_Range, _Tp>;
+
+  //! @brief Helper to check whether a different async_buffer still satisfies all properties of this one
+  template <class... _OtherProperties>
+  static constexpr bool __properties_match =
+    !_CCCL_TRAIT(_CUDA_VSTD::is_same,
+                 _CUDA_VSTD::__make_type_set<_Properties...>,
+                 _CUDA_VSTD::__make_type_set<_OtherProperties...>)
+    && _CUDA_VSTD::__type_set_contains_v<_CUDA_VSTD::__make_type_set<_OtherProperties...>, _Properties...>;
+
+  //! @brief Helper to determine what cudaMemcpyKind we need to copy data from another async_buffer with different
+  template <class... _OtherProperties>
+  static constexpr cudaMemcpyKind __transfer_kind =
+    __select_execution_space<_OtherProperties...> == _ExecutionSpace::__host
+      ? (__is_host_only ? cudaMemcpyHostToHost : cudaMemcpyHostToDevice)
+      : (__is_host_only ? cudaMemcpyDeviceToHost : cudaMemcpyDeviceToDevice);
```

Comment:
why is this needed since we have `__detect_transfer_kind`?

Reply:
Some compiler had issues if this isn't pulled into a separate variable